### PR TITLE
[Experiment] System Memory support

### PR DESCRIPTION
**Pull request by Therzok**
Using this PR as an exploration means to integrating System.Memory (Memory<T>, Span<T>, ReadOnlySpan<T>) and seeing where APIs could benefit from their usage.

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/668945

Wiki: https://github.com/mono/monodevelop/wiki/Porting-code-to-System.Memory

### CODE REVIEW

**Comment by Therzok** [comment]
@KirillOsenkov do these compat failures look sane?

```
2018-08-21T10:05:55.0725920Z   These actual lines are new:
2018-08-21T10:05:55.0792620Z   Could not load file or assembly 'System.IO.Compression.FileSystem, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089' or one of its dependencies
2018-08-21T10:05:55.0912550Z   In assembly 'netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51': unable to resolve reference to 'System.IO.Compression.FileSystem, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T10:05:55.1033760Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve member reference 'System.Boolean System.Net.ComNetOS::IsWin7orLater' from assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T10:05:55.1155700Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve member reference 'System.Boolean System.Net.Logging::get_On()' from assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T10:05:55.1278080Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve member reference 'System.Diagnostics.TraceSource System.Net.Logging::get_Http()' from assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T10:05:55.1401930Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve member reference 'System.Net.Mail.MailAddress System.Net.Mail.MailAddressParser::ParseAddress(System.String)' from assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T10:05:55.1525470Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve member reference 'System.Net.WebPermission System.Net.ExceptionHelper::WebPermissionUnrestricted' from assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T10:05:55.1650780Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve member reference 'System.Security.Cryptography.X509Certificates.X509CertificateCollection System.Net.UnsafeNclNativeMethods/NativePKI::FindClientCertificates()' from assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T10:05:55.1776150Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve member reference 'System.Void System.Net.HttpWebRequest::.ctor(System.Uri,System.Boolean,System.String,System.Action`1<System.IO.Stream>)' from assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T10:05:55.1901710Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve member reference 'System.Void System.Net.Logging::Associate(System.Diagnostics.TraceSource,System.Object,System.Object)' from assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T10:05:55.2035480Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve member reference 'System.Void System.Net.ServicePointManager::CloseConnectionGroups(System.String)' from assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T10:05:55.2155530Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve type reference 'System.Net.ComNetOS' in 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T10:05:55.2272470Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve type reference 'System.Net.Mail.MailAddressParser' in 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T10:05:55.2389230Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve type reference 'System.Net.UnsafeNclNativeMethods/NativePKI' in 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T10:05:55.2510640Z   In assembly 'Xamarin.Mac, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065': Unable to resolve member reference 'System.Net.Http.Headers.HttpHeaderKind System.Net.Http.Headers.HttpHeaders::GetKnownHeaderKind(System.String)' from assembly 'System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
2018-08-21T10:05:55.2633150Z   In assembly 'Xamarin.Mac, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065': Unable to resolve type reference 'System.Net.Http.Headers.HttpHeaderKind' in 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
```

**Comment by Therzok** [comment]
`netstandard.dll` comes from - `/reference:/Library/Frameworks/Mono.framework/Versions/5.14.0/lib/mono/xbuild/Microsoft/Microsoft.NET.Build.Extensions/net461/lib/netstandard.dll `

Seems lik a bug in msbuild?

**Comment by Therzok** [comment]
```
018-08-21T15:05:14.4042570Z   These actual lines are new:
2018-08-21T15:05:14.4110280Z   In VisualStudio.exe.config: couldn't find assembly 'System.Runtime.InteropServices.RuntimeInformation' with version 4.0.2.0. Found versions: 4.0.0.0
2018-08-21T15:05:14.4228390Z   In VisualStudio.exe.config: couldn't find assembly 'System.IO.FileSystem' with version 4.0.3.0. Found versions: 4.0.2.0
2018-08-21T15:05:14.4346910Z   In VisualStudio.exe.config: couldn't find assembly 'System.Security.Cryptography.Primitives' with version 4.0.2.0. Found versions: 4.0.1.0
2018-08-21T15:05:14.4465650Z   In VisualStudio.exe.config: couldn't find assembly 'System.IO.FileSystem.Primitives' with version 4.0.3.0. Found versions: 4.0.2.0
2018-08-21T15:05:14.4578820Z   In VisualStudio.exe.config: couldn't find assembly 'System.Diagnostics.FileVersionInfo' with version 4.0.2.0. Found versions: 4.0.1.0
2018-08-21T15:05:14.4691770Z   In VisualStudio.exe.config: couldn't find assembly 'System.Threading.Thread' with version 4.0.2.0. Found versions: 4.0.1.0
2018-08-21T15:05:14.4809500Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve member reference 'System.Boolean System.Net.ComNetOS::IsWin7orLater' from assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T15:05:14.4936100Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve member reference 'System.Boolean System.Net.Logging::get_On()' from assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T15:05:14.5057950Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve member reference 'System.Diagnostics.TraceSource System.Net.Logging::get_Http()' from assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T15:05:14.5182340Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve member reference 'System.Net.Mail.MailAddress System.Net.Mail.MailAddressParser::ParseAddress(System.String)' from assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T15:05:14.5305530Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve member reference 'System.Net.WebPermission System.Net.ExceptionHelper::WebPermissionUnrestricted' from assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T15:05:14.5429330Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve member reference 'System.Security.Cryptography.X509Certificates.X509CertificateCollection System.Net.UnsafeNclNativeMethods/NativePKI::FindClientCertificates()' from assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T15:05:14.5555440Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve member reference 'System.Void System.Net.HttpWebRequest::.ctor(System.Uri,System.Boolean,System.String,System.Action`1<System.IO.Stream>)' from assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T15:05:14.5679850Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve member reference 'System.Void System.Net.Logging::Associate(System.Diagnostics.TraceSource,System.Object,System.Object)' from assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T15:05:14.5803390Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve member reference 'System.Void System.Net.ServicePointManager::CloseConnectionGroups(System.String)' from assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T15:05:14.5925500Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve type reference 'System.Net.ComNetOS' in 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T15:05:14.6046100Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve type reference 'System.Net.Mail.MailAddressParser' in 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T15:05:14.6167690Z   In assembly 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a': Unable to resolve type reference 'System.Net.UnsafeNclNativeMethods/NativePKI' in 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
2018-08-21T15:05:14.6293590Z   In assembly 'Xamarin.Mac, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065': Unable to resolve member reference 'System.Net.Http.Headers.HttpHeaderKind System.Net.Http.Headers.HttpHeaders::GetKnownHeaderKind(System.String)' from assembly 'System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
2018-08-21T15:05:14.6415100Z   In assembly 'Xamarin.Mac, Version=0.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065': Unable to resolve type reference 'System.Net.Http.Headers.HttpHeaderKind' in 'System.Net.Http, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
```

**Comment by Therzok** [comment]
@slluis @mrward ready for review!

**Comment by Therzok** [comment]
Core failure is legit. I'll take a look.

**Review by KirillOsenkov — APPROVED** [review]

**Comment by KirillOsenkov** [comment]
Looks good. Do you have any perf numbers after spanification?

**Comment by Therzok** [comment]
> I wanted to avoid ImageCatalog and Imaging - did you need it for something?

NuGet added them.

> Looks good. Do you have any perf numbers after spanification?

I plan to get some numbers by End of Week.

File: main/external/fsharpbinding/MonoDevelop.FSharp.Tests/MonoDevelop.FSharp.Tests.fsproj [hunk 1]
```diff
@@ -314,10 +314,10 @@
     </When>
   </Choose>
   <Choose>
-    <When Condition="$(TargetFrameworkIdentifier) == '.NETFramework' And $(TargetFrameworkVersion) == 'v4.6.1'">
+    <When Condition="$(TargetFrameworkIdentifier) == '.NETFramework' And $(TargetFrameworkVersion) == 'v4.7.1'">
       <ItemGroup>
         <Reference Include="System.ValueTuple">
-          <HintPath>..\packages\System.ValueTuple\lib\net461\System.ValueTuple.dll</HintPath>
+          <HintPath>..\packages\System.ValueTuple\lib\net47\System.ValueTuple.dll</HintPath>
```
Comment (decriptor) [inline main/external/fsharpbinding/MonoDevelop.FSharp.Tests/MonoDevelop.FSharp.Tests.fsproj]: Shouldn't this be `net471`?

File: main/external/fsharpbinding/MonoDevelop.FSharpBinding/MonoDevelop.FSharp.fsproj [hunk 2]
```diff
@@ -384,10 +384,10 @@
     </When>
   </Choose>
   <Choose>
-    <When Condition="$(TargetFrameworkIdentifier) == '.NETFramework' And $(TargetFrameworkVersion) == 'v4.6.1'">
+    <When Condition="$(TargetFrameworkIdentifier) == '.NETFramework' And $(TargetFrameworkVersion) == 'v4.7.1'">
       <ItemGroup>
         <Reference Include="System.ValueTuple">
-          <HintPath>..\packages\System.ValueTuple\lib\net461\System.ValueTuple.dll</HintPath>
+          <HintPath>..\packages\System.ValueTuple\lib\net47\System.ValueTuple.dll</HintPath>
```
Comment (decriptor) [inline main/external/fsharpbinding/MonoDevelop.FSharpBinding/MonoDevelop.FSharp.fsproj]: `net471`?

File: main/external/fsharpbinding/MonoDevelop.FSharp.Tests/MonoDevelop.FSharp.Tests.fsproj (hunk 1 above)
Comment (decriptor) [reply main/external/fsharpbinding/MonoDevelop.FSharp.Tests/MonoDevelop.FSharp.Tests.fsproj]: Oh, this is coming from the nuget package.

File: main/external/fsharpbinding/MonoDevelop.FSharpBinding/MonoDevelop.FSharp.fsproj (hunk 2 above)
Comment (decriptor) [reply main/external/fsharpbinding/MonoDevelop.FSharpBinding/MonoDevelop.FSharp.fsproj]: Didn't realize this was from the nuget package

File: main/external/fsharpbinding/MonoDevelop.FSharp.Tests/MonoDevelop.FSharp.Tests.fsproj (hunk 1 above)
Comment (Therzok) [reply main/external/fsharpbinding/MonoDevelop.FSharp.Tests/MonoDevelop.FSharp.Tests.fsproj]: I'm not sure we even need it anymore in .net 471, but still, keeping it for sanity

**Review by mrward — COMMENTED** [review]
Code looks good to me. Do we have any perf numbers comparing before and after? I think Kirill already asked for this.

**Comment by Therzok** [comment]
Memory traffic for Snapshot #1 - master

Type, Allocated bytes, Allocated objects, Collected bytes, Collected objects
System.String, 1174266636, 15253024, 1091545080, 14592924

Memory traffic for Snapshot #2 - system-memory

Type, Allocated bytes, Allocated objects, Collected bytes, Collected objects
System.String, 1138190296, 16329918, 1094203580, 15963878

**Comment by Therzok** [comment]
A gain of around 30MB of allocations.

**Comment by Therzok** [comment]
All these strings will go down once we start switching more and more to ROSpan and only requiring a string when actually needed

**Review by slluis — APPROVED** [review]

**Pull request merged**